### DISCUSSION TITLE
Welcome to Pythonocc-nodes-for-Ryven Discussions!

### DISCUSSION THREAD

**Discussion by Tanneguydv** [Tanneguydv/Pythonocc-nodes-for-Ryven#2] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by leon-thomm** [comment]
### Ryven update

I've been very quiet here over the past few months, which is mainly because I was still working a lot on Ryven itself. I am doing the last tweaks for Ryven 3.1 now which is going to be a major upgrade also for this project. A few exciting technical points affecting this project:

- Ryven itself is going to be a pip package, and I am looking into adding configuration options such that it can even be loaded as a plugin from python. Notice that most of the common CAD / 3D modeling software out there has support for python plugins and often even Qt, so these nodes could be used *inside* existing 3D software.
- I improved the backend framework `ryvencore` a lot, it also gets its own repository now. When running flows made with Ryven directly on `ryvencore` (without frontend), the overhead is usually negligible. However, if ~10%-20% more speed is actually important, it will probably even be possible to add Cython support to `ryvencore`.
- You probably noticed that the flows can quickly become very slow; when you change input parameters it can take quite a bit of time until the new model is fully generated, and the viewport updates many times during this process. That is because the graphs are usually directed trees (and therefore acyclic) but they can include 'diamonds', which is a quite common pattern. For these cases I added an optimised flow execution mode which analyses the graph and makes sure that, during a computation, every node propagates its output values only once, which makes this run a lot faster.

May I also report that many people I talked to about these nodes were quite fascinated and see a lot of potential also in areas that might not be super obvious (like for programming 3D-printing-milling-hybrid systems etc.). I am super excited and hope this project really gets going with the new version.

Cheers.

**Reply by WillAdams** [reply]
Late to the party, but that's what I'm here for --- basically, for a long while I've been wanting:

 - a 3D block or node editing or flowgram system (currently using BlockSCAD)
 - the ability to write out G-Code as a model is done using 3D representations of toolpaths

Ryven w/ support for Python-OCC looks perfect --- working on installing it now, and will be writing up my results at:

https://community.carbide3d.com/t/ryven-dataflow-editor-for-python-gaining-support-for-3d/39523

**Reply by Tanneguydv** [reply]
![torus_gcode](https://user-images.githubusercontent.com/81742654/149761796-cfd5bb85-983d-403b-8589-3ec05fb74be9.jpg)

I'vre written a quick node to create gcode, could be a base for further work

**Reply by WillAdams** [reply]
Unfortunately, that file wouldn't load for me, I get:

Traceback (most recent call last):
  File "C:\Users\willa\miniconda3\envs\pyoccenv\lib\site-packages\ryven\gui\startup_dialog\StartupDialog.py", line 99, in load_example_project_button_clicked
    self.open_project(base_dir=abs_path_from_package_dir('examples_projects'))
  File "C:\Users\willa\miniconda3\envs\pyoccenv\lib\site-packages\ryven\gui\startup_dialog\StartupDialog.py", line 122, in open_project
    project_str = f.read()
  File "C:\Users\willa\miniconda3\envs\pyoccenv\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x90 in position 161607: character maps to <undefined>

**Comment by Tanneguydv** [comment]
Reaaly good news, thank you!

`May I also report that many people I talked to about these nodes were quite fascinated and see a lot of potential also in areas that might not be super obvious (like for programming 3D-printing-milling-hybrid systems etc.).`
Indeed that's the main problematics that we are adressing with this kind of softwares : to quickly generate algorithms of toolpath, CAD analysis, 3D printing strategies, architectur&design, etc. so coupling it with tons of python libraries would be very powerful

**Reply by WillAdams** [reply]
Now that I've finally gotten over the hurdle of installation, some thoughts:

 - consider looking at other projects which are similar --- BlockSCAD is the one I use most, and its ability to define modules makes for a very expressive interface
 - I'm hazy on how variables work --- one thing which I found really powerful was dividing them into parameters which were exposed in an interface for the user to interact with (think like to the Customizer in OpenSCAD) internal variables and so forth which were used for further adjustment, and then variables used for calculations
 - wondering how aesthetic the node layout for a design can be --- would it be possible to make it so that a logical and bug-free program is naturally aesthetically pleasing? Have you read Herman Hesse's _The Glass Bead Game_? Think in terms of it --- an arrangement which is logically inevitable and aesthetically beautiful
 - another tool to look at might be Full Control G-Code Designer --- currently it's an Excel spreadsheet: https://fullcontrolgcode.com/ but the next release is supposed to be an opensource Python program

Thank you for your patience and for making this wonderful tool! Really looking forward to working w/ it.
